[new release] coq-serapi (8.16.0+0.16.0)

ejgallego opened the pull request:
Serialization library and protocol for machine interaction with the Coq proof assistant

- Project page: <a href="https://github.com/ejgallego/coq-serapi">https://github.com/ejgallego/coq-serapi</a>
- Documentation: <a href="https://ejgallego.github.io/coq-serapi/">https://ejgallego.github.io/coq-serapi/</a>

##### CHANGES:

 - [serapi] (!) support for Coq 8.16, see upstream changes and SerAPI
            test-suite changes for more information. Remarkable
            changes are:
            - kernel terms are serialized a bit differently now due to
              KerName being used in more places upstream. Some
              internal structures also changes in kernel's env, so be
              attentive if you are depending on them.
            - plugin loading is adapted for 8.16 findlib loading
              method
            (@ejgallego)
 - [deps]   Require cmdliner >= 1.1.0 (@ejgallego)
 - [deps]   Support Jane Street libraries v0.15.0 (@ejgallego)
 - [serapi] New query `Objects` to dump Coq's libobject (@ejgallego)
 - [serlib] Much improved yojson / json support (@ejgallego)
 - [serlib] Coq AST now supports ppx_hash intf (ignoring locations by default)
            (@ejgallego)
 - [serlib] Coq AST now supports ppx_compare intf (ignoring locations by default)
            (@ejgallego)
 - [serlib] Large refactoring on Serlib, using functors, see serlib/README.md
            (@ejgallego)
 - [serapi] (!) Query `Proofs` has changed type and will now return the
            partial terms under construction (ejgallego/coq-serapi#271 , fixes ejgallego/coq-serapi#270, @ejgallego)

ejgallego commented:
What is the package masking `Result` module from Stdlib ? I read about this somewhere. Sorry, but this seems a PITA.

kit-ty-kate commented on packages/coq-serapi/coq-serapi.8.16.0+0.16.0/opam:
```diff
@@ -0,0 +1,50 @@
+  "ppx_hash"            {           >= "v0.13.0" & < "v0.16" }
+  "yojson"              {           >= "1.7.0"               }
+  "ppx_deriving_yojson" {           >= "3.4"                 }
+]
```
```suggestion
]
conflicts: [
  "result" {< "1.5"}
]
```

ejgallego commented:
Thanks a lot for the fix, should I port this to upstream opam file?

kit-ty-kate commented:
Yes please! ^^

kit-ty-kate commented:
Do you have any idea about what's happening on x86_32?
```
# (cd _build/default && /home/opam/.opam/4.14/bin/ocamlopt.opt -w -40 -rectypes -g -ppx '.ppx/341932708445810a638020e478ebc4f1/ppx.exe --as-ppx --cookie '\''library-name="serlib"'\''' -I serlib/.serlib.objs/byte -I serlib/.serlib.objs/native -I /home/opam/.opam/4.14/lib/base -I /home/opam/.opam/4.14/lib/base/base_internalhash_types -I /home/opam/.opam/4.14/lib/base/caml -I /home/opam/.opam/4.14/lib/base/shadow_stdlib -I /home/opam/.opam/4.14/lib/coq-core/boot -I /home/opam/.opam/4.14/lib/coq-core/clib -I /home/opam/.opam/4.14/lib/coq-core/config -I /home/opam/.opam/4.14/lib/coq-core/engine -I /home/opam/.opam/4.14/lib/coq-core/gramlib -I /home/opam/.opam/4.14/lib/coq-core/interp -I /home/opam/.opam/4.14/lib/coq-core/kernel -I /home/opam/.opam/4.14/lib/coq-core/lib -I /home/opam/.opam/4.14/lib/coq-core/library -I /home/opam/.opam/4.14/lib/coq-core/parsing -I /home/opam/.opam/4.14/lib/coq-core/pretyping -I /home/opam/.opam/4.14/lib/coq-core/printing -I /home/opam/.opam/4.14/lib/coq-core/proofs -I /home/opam/.opam/4.14/lib/coq-core/stm -I /home/opam/.opam/4.14/lib/coq-core/sysinit -I /home/opam/.opam/4.14/lib/coq-core/tactics -I /home/opam/.opam/4.14/lib/coq-core/vernac -I /home/opam/.opam/4.14/lib/coq-core/vm -I /home/opam/.opam/4.14/lib/findlib -I /home/opam/.opam/4.14/lib/ocaml/threads -I /home/opam/.opam/4.14/lib/parsexp -I /home/opam/.opam/4.14/lib/ppx_compare/runtime-lib -I /home/opam/.opam/4.14/lib/ppx_deriving/runtime -I /home/opam/.opam/4.14/lib/ppx_deriving_yojson/runtime -I /home/opam/.opam/4.14/lib/ppx_hash/runtime-lib -I /home/opam/.opam/4.14/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.14/lib/result -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/yojson -I /home/opam/.opam/4.14/lib/zarith -intf-suffix .ml -no-alias-deps -open Serlib -o serlib/.serlib.objs/native/serlib__Ser_glob_term.cmx -c -impl serlib/ser_glob_term.ml)
# /opam-tmp/build_da913c_dune/camlasm42ade0.s: Assembler messages:
# /opam-tmp/build_da913c_dune/camlasm42ade0.s:11500: Error: operand size mismatch for `movzb'
# File "serlib/ser_glob_term.ml", line 1:
# Error: Assembler error, input left in file /opam-tmp/build_da913c_dune/camlasm42ade0.s
```

ejgallego commented:
> Do you have any idea about what's happening on x86_32?

Interesting, no idea, but SerAPI uses some tricky stuff. I happen to have a 32 bit switch for jscoq around, let me try locally.

ejgallego commented:
Works fine here, Ubuntu 22.04, tho 32 bit support toolchain is (IMVHO) not in the very best shape from what I see, quite a few warnings with recent compilers.

kit-ty-kate commented on packages/coq-serapi/coq-serapi.8.16.0+0.16.0/opam:
```diff
@@ -0,0 +1,53 @@
+  "ocamlfind"           {           >= "1.8.0"               }
+  "sexplib"             {           >= "v0.13.0"             }
+  "dune"                {           >= "2.0.1"               }
+  "ppx_import"          { build   & >= "1.5-3"               }
```
```suggestion
  "ppx_import"          { build   & >= "1.5-3" & < "2.0"     }
```

kit-ty-kate commented:
Ok let's ignore that failure then. I've added a constraint to ppx_import since the syntax will change entirely.

ejgallego commented:
> Ok let's ignore that failure then. I've added a constraint to ppx_import since the syntax will change entirely.

Oh indeed :) thanks for the reminder, I've updated the upstream opam.

> Upper-bound constraints on the janestreet packages are not necessary and makes it harder to upgrade the ecosystem when a new release hits.

Unfortunately jane street packages are not API compatible for new upstream releases, these were added because serapi CI broke. What do you suggest?

kit-ty-kate commented:
> these were added because serapi CI broke on new JS releases, and I'd rather avoid that.

In my opinion, the role of a CI is to break when an action is needed, you can always re-add the constraints when needed if you need to do something before moving to a new api.

ejgallego commented:
> In my opinion, the role of a CI is to break when an action is needed, you can always re-add the constraints when needed if you need to do something before moving to a new api.

I want my CI to be a bit more deterministic I'm afraid. I don't want to have to deal with this kind of problem in an unplanned manner.

The key point is that the constraints _may be or may be not be necessary_, so in this case I prefer to err in the safe side.

Experience has shown me that usually _there are necessary_.

kit-ty-kate commented:
In that case we'll have to remove them manually for each new releases of coq-serapi. If you can remember to add a commit next time on top of the PR it would be very appreciated, but if you don't that's ok we'll probably do it for you (though I can't promise that I will remember this discussion next time, and we might end up having the same discussion again :sweat_smile:)

ejgallego commented:
Yes we already had this discussion, hmm, let me see if I can switch my CI to use an opam-lockfile, so the opam file is correct for you folks upstream.

kit-ty-kate commented:
Thanks a lot! In the meantime we can merge this.

ejgallego commented:
> Thanks a lot! In the meantime we can merge this.

Thanks to you for all the feedback and help, I think upstream opam file should be fine now.
